src/lib/OpenEXR/ImfContextInit.cpp

```diff
@@ -0,0 +1,213 @@
+//
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) Contributors to the OpenEXR Project.
+//
+
+#include "ImfContextInit.h"
+#include "ImfIO.h"
+
+#include <cinttypes>
+#include <climits>
+#include <cstring>
+#include <mutex>
+
+OPENEXR_IMF_INTERNAL_NAMESPACE_SOURCE_ENTER
+
+struct istream_holder
+{
+    istream_holder (IStream* s) : _stream (s)
+    {
+    }
+
+#if ILMTHREAD_THREADING_ENABLED
+    std::mutex _mx;
+#endif
+    IStream* _stream;
+};
+
+static int64_t
+istream_read (
+    exr_const_context_t         ctxt,
+    void*                       userdata,
+    void*                       buffer,
+    uint64_t                    sz,
+    uint64_t                    offset,
+    exr_stream_error_func_ptr_t error_cb)
+{
+    istream_holder* ih    = static_cast<istream_holder*> (userdata);
+    IStream*        s     = ih->_stream;
+
+    if (sz > INT_MAX)
+    {
+        error_cb (
+            ctxt,
+            EXR_ERR_READ_IO,
+            "Stream interface request to read block too large");
+        return -1;
+    }
+
+#if ILMTHREAD_THREADING_ENABLED
+    std::lock_guard<std::mutex> lk{ih->_mx};
+#endif
+
+    int64_t         nread = s->tellg ();
+    try
+    {
+        if (offset != static_cast<size_t> (nread))
+        {
+            s->seekg (offset);
+            nread = s->tellg ();
+            if (offset != static_cast<size_t> (nread))
+            {
+                error_cb (
+                    ctxt,
+                    EXR_ERR_READ_IO,
+                    "Unable to seek to desired offset %" PRIu64,
+                    offset);
+                return -1;
+            }
+        }
+
+        try
+        {
+            s->read (static_cast<char*> (buffer), static_cast<int> (sz));
+        }
+        catch (...)
+        {
+            // bah, there could be two reasons for this, one is a
+            // legitimate error, the other is a read past the end of file
+            // (i.e. the core library tries to read a 4k block when
+            // parsing the header), let's let the core deal with that and
+            // clear errors
+            ih->_stream->clear ();
+            //error_cb (
```

Comment:
will this be re-enabled for check in?

Reply:
we need to add new api to the stream class to do this properly, for a small file, there very well may be requests that (unknowingly) request more data than what is in the file, this used to throw, but should just return how much it read...